### ISSUE TITLE
'Bad record mac' error still an issue with correct 'gun' dependency

### ISSUE THREAD

**Issue by GoNZooo** [Kraigie/nostrum#29]
In the README it's stated that you have to fetch the 'dd1bfe4d6f9fb277781d922aa8bbb5648b3e6756' ref of gun to make the library work with OTP 20.

Doing so and then using the master branch (no ref/tag specified) of nostrum seems to have the same issue as without the supposed fix.

I'm running Elixir 1.5.2, OTP 20 and the specified dependency specifications. Any ideas on why this still seems to be an issue?

Here is the error message, for posterity's sake:

```elixir
==> pomoshtnik_discord
Compiling 2 files (.ex)
Generated pomoshtnik_discord app

20:35:18.626 [error] SSL: :cipher: ssl_alert.erl:88:Fatal error: bad record mac


20:35:18.629 [error] SSL: :cipher: ssl_alert.erl:88:Fatal error: bad record mac


20:35:18.652 [error] SSL: :cipher: ssl_alert.erl:88:Fatal error: bad record mac


20:35:18.671 [error] SSL: :cipher: ssl_alert.erl:88:Fatal error: bad record mac


20:35:18.695 [error] SSL: :cipher: ssl_alert.erl:88:Fatal error: bad record mac


20:35:18.725 [error] SSL: :cipher: ssl_alert.erl:88:Fatal error: bad record mac


=INFO REPORT==== 14-Oct-2017::20:35:18 ===
    application: logger
    exited: stopped
    type: temporary
** (Mix) Could not start application nostrum: Nostrum.start(:normal, []) returned an error: shutdown: failed to start child: Nostrum.Shard.Supervisor
    ** (EXIT) shutdown: failed to start child: 0
        ** (EXIT) shutdown: failed to start child: Nostrum.Shard.Session
            ** (EXIT) an exception was raised:
                ** (ErlangError) Erlang error: {:gone, {:tls_alert, 'bad record mac'}}
                    (gun) /home/gonz/code/elixir/pomoshtnik/deps/gun/src/gun.erl:596: :gun.retry/2
                    (gun) /home/gonz/code/elixir/pomoshtnik/deps/gun/src/gun.erl:528: :gun.proc_lib_hack/5
                    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
```

Here is the lock-file:

```
%{"certifi": {:hex, :certifi, "1.0.0", "1c787a85b1855ba354f0b8920392c19aa1d06b0ee1362f9141279620a5be2039", [], [], "hexpm"},
  "cowlib": {:git, "https://github.com/ninenines/cowlib", "bd37be4d3b065600c3b76b492535e76e5d413fc1", [ref: "master"]},
  "gen_stage": {:hex, :gen_stage, "0.12.2", "e0e347cbb1ceb5f4e68a526aec4d64b54ad721f0a8b30aa9d28e0ad749419cbb", [], [], "hexpm"},
  "gun": {:git, "https://github.com/ninenines/gun.git", "dd1bfe4d6f9fb277781d922aa8bbb5648b3e6756", [ref: "dd1bfe4d6f9fb277781d922aa8bbb5648b3e6756"]},
  "hackney": {:hex, :hackney, "1.7.1", "e238c52c5df3c3b16ce613d3a51c7220a784d734879b1e231c9babd433ac1cb4", [], [{:certifi, "1.0.0", [hex: :certifi, repo: "hexpm", optional: false]}, {:idna, "4.0.0", [hex: :idna, repo: "hexpm", optional: false]}, {:metrics, "1.0.1", [hex: :metrics, repo: "hexpm", optional: false]}, {:mimerl, "1.0.2", [hex: :mimerl, repo: "hexpm", optional: false]}, {:ssl_verify_fun, "1.1.1", [hex: :ssl_verify_fun, repo: "hexpm", optional: false]}], "hexpm"},
  "httpoison": {:hex, :httpoison, "0.11.1", "d06c571274c0e77b6cc50e548db3fd7779f611fbed6681fd60a331f66c143a0b", [], [{:hackney, "~> 1.7.0", [hex: :hackney, repo: "hexpm", optional: false]}], "hexpm"},
  "idna": {:hex, :idna, "4.0.0", "10aaa9f79d0b12cf0def53038547855b91144f1bfcc0ec73494f38bb7b9c4961", [], [], "hexpm"},
  "metrics": {:hex, :metrics, "1.0.1", "25f094dea2cda98213cecc3aeff09e940299d950904393b2a29d191c346a8486", [], [], "hexpm"},
  "mimerl": {:hex, :mimerl, "1.0.2", "993f9b0e084083405ed8252b99460c4f0563e41729ab42d9074fd5e52439be88", [], [], "hexpm"},
  "nostrum": {:git, "https://github.com/Kraigie/nostrum", "06af555ea07fc09b340740eb7c10345551be8799", []},
  "poison": {:hex, :poison, "3.1.0", "d9eb636610e096f86f25d9a46f35a9facac35609a7591b3be3326e99a0484665", [], [], "hexpm"},
  "ranch": {:git, "https://github.com/ninenines/ranch", "70f9696b6c35d029ac7bce67485713aee3983917", [ref: "master"]},
  "slack": {:hex, :slack, "0.12.0", "a305f87adca043e056a86f92151d085a8c359718f9c4ca7f1d1012d67b75db8a", [], [{:httpoison, "~> 0.11", [hex: :httpoison, repo: "hexpm", optional: false]}, {:poison, "~> 3.0", [hex: :poison, repo: "hexpm", optional: false]}, {:websocket_client, "~> 1.2.4", [hex: :websocket_client, repo: "hexpm", optional: false]}], "hexpm"},
  "ssl_verify_fun": {:hex, :ssl_verify_fun, "1.1.1", "28a4d65b7f59893bc2c7de786dec1e1555bd742d336043fe644ae956c3497fbe", [], [], "hexpm"},
  "websocket_client": {:hex, :websocket_client, "1.2.4", "14ec1ca4b6d247b44ccd9a80af8f6ca98328070f6c1d52a5cb00bc9d939d63b8", [], [], "hexpm"}}
```

**Comment by Kraigie** [comment]
Hi, thanks for the report! 

What operating system are you running this on?

**Comment by GoNZooo** [comment]
This is running on ArchLinux, fully updated. The OpenSSL version is 1.1.0.f-2 and I've actually tried juggling both Erlang and Elixir versions to no avail. I'm starting to think this is Erlang-specific with this setup, because another discord library also gave the same error.

**Comment by Kraigie** [comment]
Someone else has reported that this also occurred for them on Arch. I'll dig around a bit and see if I can find anything.

**Comment by skwerlman** [comment]
any news on this?

**Comment by Kraigie** [comment]
Sorry for the late reply, it looks like this is an issue unique to arch.

For now it looks like the only option is to downgrade. It looks like openssl 1.1.0 support was [added in crypto 4.0](http://erlang.org/doc/apps/crypto/notes.html). I'm not sure where on the timeline this update fell, but [commits point towards late 2016](https://github.com/erlang/otp/commit/53aa4605b8c23ec1cb7d5a79625db00ebdd63423), meaning this looks like it should be implemented.

I'll leave this open for now, in case some more solid info is found. Feel free to close if you wish.

**Comment by skwerlman** [comment]
this is the commit the arch package is built from: https://github.com/erlang/otp/commit/a01de6873844ba510084090abec734c4166d71fa

the PKGBUILD: https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/erlang

Oddly, the PKGBUILD seems to be for erlang-20.1, while the version in the repos is erlang-20.0.1. Additionally, it looks like the package is flagged as out-of-date.

I'm gonna try building from the newer PKGBUILD, and I'll report back if I have any success.

**Comment by skwerlman** [comment]
Unfortunately, the error persists even with erlang 20.1.6...

I'm wondering now if this has something to do with the openssl versions used by arch. A few arch packages depend on `openssl-1.0`, which is currently at version `1.0.2.l-1`, so maybe that's causing erlang to trip up? I'm not sure how to debug this, since removing it would break a fair number of packages for me (including all of qt5!).

@GoNZooo are you able to remove `openssl-1.0` from your system? (or do you even have it installed?)

**Comment by Kayyow** [comment]
Same issue using Arch fully updated

**Comment by perillamint** [comment]
It is known problem of OTP on Arch (https://stackoverflow.com/questions/45245365/bad-record-mac-with-ssl-connecting-to-cloudflare-on-arch-linux). CF's SSL impl tries to connect using cipher 0xCC14 and one of TLS impl is broken on 0xCC14 so CF spits "Bad Record MAC" to us.

I have no idea about it is CF's fault or Arch's fault. Cipher suite 0xCC14 is draft anyway https://github.com/mozilla/cipherscan/issues/118

I'm currently using Docker to develop my elixir project on Arch. I also tried compiling OTP package with some options, but I failed to make it work with CF.

**Comment by skwerlman** [comment]
issue still exists with erlang 20.2-1/crypto 4.2

**Comment by HatefulMoron** [comment]
Uninstalling the elixir Arch package with `sudo pacman -R elixir` and manually building+installing the master branches of OTP and elixir solved this issue for me for the meantime.

**Comment by skwerlman** [comment]
just installing OTP from master should be enough, as it already contains a temporary fix. no need to install elixir by hand.

**Comment by snapcase** [comment]
it seems that this is no longer an issue on Arch Linux. I have it running on `elixir-1.6.2-1` without the specific gun dependency.

**Comment by skwerlman** [comment]
so the fix made it into erlang-2.0.3? if so this can probably be closed

**Comment by Kraigie** [comment]
I've confirmed with a couple people that this has been resolved with erts 9.3 so I'll be closing this now.

If you find this isn't the case or have any issues feel free to leave another comment, or contact @snapcase since he's an expert on this issue (not really, but he loves the attention)

**Comment by skwerlman** [comment]
Oddly the issue persists in `erlang 20.3.2-1`, but is fixed in `erlang-nox 20.3-1`

**Comment by PixeLInc** [comment]
Can confirm, issue exists in the latest version of erlang-nox. Downgrading back to 20.3-1 makes it work again. 
Weird that they broke it again :thinking:
